Feature/friends funtionality

Sekir0 opened the pull request:
1 - filters in mongo repo
2 - model is very simple (I think)
3 - enum "RequesrStatus"
4 - what should I do with a declined friendship request

Sekir0 committed:
Add base frend functionality

add to friend
delete friend
get list friends

Sekir0 committed:
upd

+following list
+followers list
+friends list

Sekir0 committed:
upd

extend friend document
add SearchFriendRequests endpoint

Sekir0 committed:
Merge branch 'master' into feature/friends_funtionality

Sekir0 committed:
Merge branch 'master' into feature/friends_funtionality

Sekir0 committed:
chang friend request logic

Sekir0 committed:
add ability to delete friend request

fix naming

Sekir0 committed:
add ability to accept friend requests

Sekir0 committed:
get sended  friend requests / update friend request

Sekir0 committed:
upd

change filters

Sekir0 committed:
Merge branch 'master' into feature/friends_funtionality

Sekir0 committed:
add async postfix

Sekir0 committed:
change filter for delete friend

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendFunctionalityController.cs:
```diff
@@ -0,0 +1,138 @@
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class FriendFunctionalityController : Controller
```
Controller is overkill. Use ControllerBase

Remove "functionality" from name

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendFunctionalityController.cs:
```diff
@@ -0,0 +1,138 @@
+        /// <param name="take">Take friends up to a specified position</param>
+        /// <param name="userId">Filters friends by userId</param>
+        /// <response code="200">Return friends</response>
+        [HttpGet("friends/{userId}")]
```
Controller already contains "friends"

volodymyr-borodin reviewed (commented):
- no need to store declined request
- no need to add request from user1 to user2 if it is already exists
- I believe that id field is redundant. FromUser + ToUser should be uniq and can be used as primary key
- Status redundant as well. It can be presented in the DB but what is the reason to have it in the output model?
- Instead of one update status method it will be great to have 2 different methods
- - Accept request
- - Decline request

volodymyr-borodin commented on Infrastructure/GhostNetwork.Profiles.MongoDb/MongoFriendRequestStorage.cs:
```diff
@@ -0,0 +1,132 @@
+
+        public async Task InsertFriendRequestAsync(FriendRequest friendRequest)
+        {
+            var filter = Builders<FriendRequestEntity>.Filter.Where(x => x.FromUser == friendRequest.FromUser && x.ToUser == friendRequest.ToUser)
```
Replace Where with Eq filter

Sekir0 committed:
upd

split update method to accept & decline
rename controller
now declined reqest is not stored

Sekir0 committed:
upd

Change Friend resp model
Change filters

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class FriendController : ControllerBase
```
"friendS" might be better

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        [HttpGet("{userId}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [SwaggerResponseHeader(StatusCodes.Status200OK, "X-TotalCount", "Number", "Total number of user friends")]
+        public async Task<ActionResult<IEnumerable<Friend>>> SearchFriendsAsync(
```
I'm little bit concerned about this model. It is not clear for me who is FromUser and who is ToUser. It will be better to return OR array of friends ids OR short information about friend e.g.

```json
{
  "id": "some-id",
  "fullName": "friend's name"
}
```

The same for followers and followed

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <param name="take">Take friends up to a specified position</param>
+        /// <param name="userId">Filters friends by userId</param>
+        /// <response code="200">Return friends</response>
+        [HttpGet("{userId}")]
```
I can be discussed but I think more RESTful style will be [GET] /users/{userId}/friends

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <param name="take">Take friends up to a specified position</param>
+        /// <param name="userId">Filters by userId</param>
+        /// <response code="200">Return received friend requests</response>
+        [HttpGet("followers/{userId}")]
```
It is better to have [GET] /users/{userId}/followers

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <param name="take">Take friends up to a specified position</param>
+        /// <param name="userId">Filters friends by userId</param>
+        /// <response code="200">Return sended friend requests</response>
+        [HttpGet("followed/{userId}")]
```
It is better to have [GET] /users/{userId}/followed

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// <param name="fromUser">Filters friend request by fromUser</param>
+        /// <param name="toUser">Filters friend request by toUser</param>
+        /// <response code="200">Friend request successfully created</response>
+        [HttpPost("friendrequest/{fromUser}/{toUser}")]
```
Url looks strange /friends/friendrequest/{fromUser}/{toUser}
Maybe it is better to have something like /friends/{fromUser}/{toUser}?

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// </summary>
+        /// <param name="id">Filters friend request by id</param>
+        /// <response code="200">Friend request successfully accepted</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
```
Can we use POST for approve request?

For example
- there is not request for user1 to user2 -> request will be created
- there is request for user1 to user2 -> request will be approved
- there is request for user2 to user1 -> no actions required

volodymyr-borodin commented on GhostNetwork.Profiles.Api/Controllers/FriendController.cs:
```diff
@@ -0,0 +1,142 @@
+        /// </summary>
+        /// <param name="id">Friend request id</param>
+        /// <response code="200">Friend successfully deleted =(</response>
+        [HttpDelete("{id}")]
```
Change to [Delete] /friends/{fromUser}/{toUser}

Can we combine logic of delete from friend and decline request in one endpoint?

Sekir0 replied:
Yep, we can just delete the request. I think it's be better
Thanks for review
I'll fix this all

Sekir0 committed:
Follow the REST standarts

rename friend store and service

Sekir0 committed:
Add new response model for followers and follower
